grant-project/zcash-grant-system
# Arbiter management

Closes #174.

More robust proposal arbiters. Nomination instead of straight setting by admin. I was going to roll this into milestone payout flow, but it got too big for that.

### What this does
- breaks arbiter into its own `ProposalArbiterModel`
- adds `ProposalArbiterStatus` MISSING, NOMINATED, ACCEPTED
- email contains a link to accept nomination
- profile "arbitrations" tab shows "accept"/"reject" buttons for NOMINATED state
- profile "arbitrations" tab contains a link to abdicate arbiter responsibilities
- admin proposal filters for all arbiter states added
- admin proposal detail "nominate arbiter" action item only enabled when proposal status is LIVE
- admin cannot set arbiter directly, only nominate -- which will remove existing arbiter (nominated or accepted) if present

### Testing
- you will likely have to nix your db, I didn't write migration to create/move existing arbiters over
- `flask db upgrade`
- create a proposal and go through steps necessary to make it LIVE (published)
- admin: visit the proposal detail page, click "nominate arbiter" -- select an arbiter with email you can access
- check the email, should have an "accept nomination" button, click it
- should succeed, click link to "view arbitrations"
- should show the arbitrated proposal, along with a link to "opt out", click it and confirm
- go back to admin and confirm you get the no arbiter state, nominate the user again
- this time don't use the email, instead use the profile/arbitrations tab and click "reject"
- go back to admin and once again nominate the user
- refresh the profile/arbiters page and this time click "accept", should update with ACCEPTED state 
- admin should not show and Alert, but should have an enabled "Change nomination" in the actions on the side

## Screens

### admin detail, LIVE proposal, arbiter MISSING
<img width="902" alt="screen shot 2019-02-09 at 9 23 23 pm" src="https://user-images.githubusercontent.com/11430954/52529212-c8bb5880-2cb3-11e9-95f3-f0ad41520e11.png">

### admin detail, LIVE proposal, arbiter NOMINATED
<img width="900" alt="screen shot 2019-02-09 at 9 23 49 pm" src="https://user-images.githubusercontent.com/11430954/52529214-d244c080-2cb3-11e9-9149-411ec5f77b3e.png">

### user profile/arbitrations, user nominated arbiter
<img width="829" alt="screen shot 2019-02-09 at 9 24 24 pm" src="https://user-images.githubusercontent.com/11430954/52529215-d7a20b00-2cb3-11e9-91a0-a09f54a4cf29.png">

### user email for arbiter nomination
<img width="886" alt="screen shot 2019-02-09 at 9 21 51 pm" src="https://user-images.githubusercontent.com/11430954/52529219-dc66bf00-2cb3-11e9-9fce-9cad5fb6a9e7.png">

### user profile/arbitrations, user accepted - clicked "opt-out"
<img width="855" alt="screen shot 2019-02-09 at 9 25 11 pm" src="https://user-images.githubusercontent.com/11430954/52529221-e1c40980-2cb3-11e9-8d1e-c51fc322051b.png">

### admin detail, LIVE proposal, arbiter ACCEPTED
<img width="900" alt="screen shot 2019-02-09 at 9 25 48 pm" src="https://user-images.githubusercontent.com/11430954/52529222-e5579080-2cb3-11e9-99c1-30739b757c44.png">
